bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java

```
@@ -44,14 +45,133 @@
+          Iterator<SettableFuture<Empty>> entries = futures.iterator();
+          while (entries.hasNext() && statuses.hasNext()) {
+            SettableFuture<Empty> future = entries.next();
+            Status status = statuses.next();
+            if (status.getCode() == io.grpc.Status.Code.OK.value()) {
+              future.set(Empty.getDefaultInstance());
+            } else {
+              future.setException(toException(status));
+            }
+          }
+          // TODO: better handling of these cases?
+          while (entries.hasNext()) {
+            entries.next().setException(MISSING_ENTRY_EXCEPTION);
+          }
+          if (statuses.hasNext()) {
+            int count = 0;
+            while (statuses.hasNext()) {
+              count++;
+              statuses.next();
+            }
+            throw new IllegalStateException(String.format("Got %d extra statusus", count));
+          }
+        }
+
+        protected StatusRuntimeException toException(Status status) {
+          io.grpc.Status grpcStatus = io.grpc.Status
+              .fromCodeValue(status.getCode())
+              .withDescription(status.getMessage());
+          for (Any detail : status.getDetailsList()) {
+            grpcStatus.augmentDescription(detail.toString());
+          }
+          return grpcStatus.asRuntimeException();
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+          for (SettableFuture<Empty> future : futures) {
+            future.setException(t);
+          }
+        }
+      };
+      Futures.addCallback(bulkFuture, callback);
+    }
+
+    private void mutateRowAsync() {
+      ListenableFuture<MutateRowsResponse> future = null;
+      try {
+        future = asyncExecutor.mutateRowsAsync(toRequest());
+      } catch (InterruptedException e) {
+        future = Futures.<MutateRowsResponse> immediateFailedFuture(e);
+      } finally {
+        if (future != null) {
```

If it's possible that the future could be null here it might be better to raise an Exception rather than skip the callback.